addons/binding/org.openhab.binding.vera2/src/main/java/org/openhab/binding/vera2/internal/discovery/VeraDeviceDiscoveryService.java

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.vera2.internal.discovery;
+
+import static org.openhab.binding.vera2.VeraBindingConstants.*;
+
+import java.util.List;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.vera2.controller.json.Device;
+import org.openhab.binding.vera2.controller.json.Scene;
+import org.openhab.binding.vera2.handler.VeraBridgeHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link VeraDeviceDiscoveryService} is responsible for device discovery.
+ *
+ * @author Dmitriy Ponomarev
+ */
+public class VeraDeviceDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private static final int SEARCH_TIME = 60;
+    private static final int INITIAL_DELAY = 15;
+    private static final int SCAN_INTERVAL = 240;
+
+    private VeraBridgeHandler mBridgeHandler;
+    private VeraDeviceScan mVeraDeviceScanningRunnable;
+    private ScheduledFuture<?> mVeraDeviceScanningJob;
+
+    public VeraDeviceDiscoveryService(VeraBridgeHandler bridgeHandler) {
+        super(SUPPORTED_DEVICE_THING_TYPES_UIDS, SEARCH_TIME);
+        logger.debug("Initializing VeraDeviceDiscoveryService");
+        mBridgeHandler = bridgeHandler;
+        mVeraDeviceScanningRunnable = new VeraDeviceScan();
+        activate(null);
+    }
+
+    private void scan() {
+        logger.debug("Starting scan on Vera Controller {}", mBridgeHandler.getThing().getUID());
+
+        if (mBridgeHandler == null || !mBridgeHandler.getThing().getStatus().equals(ThingStatus.ONLINE)) {
+            logger.debug("Vera bridge handler not found or not ONLINE.");
+            return;
+        }
+
+        final ThingUID bridgeUID = mBridgeHandler.getThing().getUID();
+
+        List<Device> deviceList = mBridgeHandler.getData().devices;
+        for (Device device : deviceList) {
+            if (device.category.equals("0")) {

```

Comment:
Position literals first in String comparisons - that way if the String is null you won't get a NullPointerException, it'll just return false. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)